In [131]:
import pandas as pd
import numpy as np

# Importing Atlantic Hurricane data

In [110]:
atlantic = pd.read_csv('/Users/danielbrechner/hurrdat_co2/assets/atlantic.csv')
atlantic.head()

ID                 Name      Date  Time Event Status Latitude  \
0  AL011851              UNNAMED  18510625     0           HU    28.0N   
1  AL011851              UNNAMED  18510625   600           HU    28.0N   
2  AL011851              UNNAMED  18510625  1200           HU    28.0N   
3  AL011851              UNNAMED  18510625  1800           HU    28.1N   
4  AL011851              UNNAMED  18510625  2100     L     HU    28.2N   

  Longitude  Maximum Wind  Minimum Pressure  ...  Low Wind SW  Low Wind NW  \
0     94.8W            80              -999  ...         -999         -999   
1     95.4W            80              -999  ...         -999         -999   
2     96.0W            80              -999  ...         -999         -999   
3     96.5W            80              -999  ...         -999         -999   
4     96.8W            80              -999  ...         -999         -999   

   Moderate Wind NE  Moderate Wind SE  Moderate Wind SW  Moderate Wind NW  \
0              -999              -999              -999              -999   
1              -999              -999              -999              -999   
2              -999              -999              -999              -999   
3              -999              -999              -999              -999   
4              -999              -999              -999              -999   

   High Wind NE  High Wind SE  High Wind SW  High Wind NW  
0          -999          -999          -999          -999  
1          -999          -999          -999          -999  
2          -999          -999          -999          -999  
3          -999          -999          -999          -999  
4          -999          -999          -999          -999  

[5 rows x 22 columns]

# Data Preprocessing

In [111]:
atlantic.shape

(49105, 22)

In [112]:
atlantic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49105 entries, 0 to 49104
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                49105 non-null  object
 1   Name              49105 non-null  object
 2   Date              49105 non-null  int64 
 3   Time              49105 non-null  int64 
 4   Event             49105 non-null  object
 5   Status            49105 non-null  object
 6   Latitude          49105 non-null  object
 7   Longitude         49105 non-null  object
 8   Maximum Wind      49105 non-null  int64 
 9   Minimum Pressure  49105 non-null  int64 
 10  Low Wind NE       49105 non-null  int64 
 11  Low Wind SE       49105 non-null  int64 
 12  Low Wind SW       49105 non-null  int64 
 13  Low Wind NW       49105 non-null  int64 
 14  Moderate Wind NE  49105 non-null  int64 
 15  Moderate Wind SE  49105 non-null  int64 
 16  Moderate Wind SW  49105 non-null  int64 
 17  Moderate Win

In [113]:
atlantic.describe()

Date          Time  Maximum Wind  Minimum Pressure  \
count  4.910500e+04  49105.000000  49105.000000      49105.000000   
mean   1.949802e+07    910.125975     52.005091       -251.406496   
std    4.461850e+05    671.043363     27.681902        964.313907   
min    1.851062e+07      0.000000    -99.000000       -999.000000   
25%    1.911110e+07    600.000000     35.000000       -999.000000   
50%    1.956093e+07   1200.000000     45.000000       -999.000000   
75%    1.989081e+07   1800.000000     70.000000        990.000000   
max    2.015111e+07   2330.000000    165.000000       1024.000000   

        Low Wind NE   Low Wind SE   Low Wind SW   Low Wind NW  \
count  49105.000000  49105.000000  49105.000000  49105.000000   
mean    -868.671031   -869.315772   -872.676428   -871.409245   
std      353.299666    351.548910    342.153768    345.628984   
min     -999.000000   -999.000000   -999.000000   -999.000000   
25%     -999.000000   -999.000000   -999.000000   -999.000000   
50%     -999.000000   -999.000000   -999.000000   -999.000000   
75%     -999.000000   -999.000000   -999.000000   -999.000000   
max      710.000000    600.000000    640.000000    530.000000   

       Moderate Wind NE  Moderate Wind SE  Moderate Wind SW  Moderate Wind NW  \
count      49105.000000      49105.000000      49105.000000      49105.000000   
mean        -875.570940       -875.765319       -876.682028       -876.323205   
std          333.651396        333.133376        330.526085        331.535176   
min         -999.000000       -999.000000       -999.000000       -999.000000   
25%         -999.000000       -999.000000       -999.000000       -999.000000   
50%         -999.000000       -999.000000       -999.000000       -999.000000   
75%         -999.000000       -999.000000       -999.000000       -999.000000   
max          360.000000        300.000000        330.000000        360.000000   

       High Wind NE  High Wind SE  High Wind SW  High Wind NW  
count  49105.000000  49105.000000  49105.000000  49105.000000  
mean    -877.564321   -877.655045   -877.923552   -877.786295  
std      328.027275    327.774794    327.021430    327.408253  
min     -999.000000   -999.000000   -999.000000   -999.000000  
25%     -999.000000   -999.000000   -999.000000   -999.000000  
50%     -999.000000   -999.000000   -999.000000   -999.000000  
75%     -999.000000   -999.000000   -999.000000   -999.000000  
max      180.000000    250.000000    150.000000    180.000000

In [114]:
atlantic.isnull().values.any() 

False

No null values are found yet -999 is a nonsensical value for wind values. Lets drop these values from the dataset.

Dropping unnecessary columns

In [115]:
atlantic.drop(['ID', 'Time', 'Minimum Pressure', 'Event'], axis = 1, inplace= True)

In [130]:
#Finding the negative values where column dtype is int64 and printing the sum of those values
num_cols = atlantic.select_dtypes(include=['int64']).columns
neg_counts = {}
for col in num_cols:
    neg_counts[col] = len(atlantic[atlantic[col] < 0])
neg_counts


{'Maximum Wind': 338,
 'Low Wind NE': 43184,
 'Low Wind SE': 43184,
 'Low Wind SW': 43184,
 'Low Wind NW': 43184,
 'Moderate Wind NE': 43184,
 'Moderate Wind SE': 43184,
 'Moderate Wind SW': 43184,
 'Moderate Wind NW': 43184,
 'High Wind NE': 43184,
 'High Wind SE': 43184,
 'High Wind SW': 43184,
 'High Wind NW': 43184,
 'Year': 0}

In [133]:
atlantic[num_cols] = atlantic[num_cols].mask(atlantic[num_cols] < 0, np.nan)

In [134]:
sum_negative = (atlantic.iloc[:, 5:20] < 0).sum()
print(sum_negative)

Maximum Wind        0
Low Wind NE         0
Low Wind SE         0
Low Wind SW         0
Low Wind NW         0
Moderate Wind NE    0
Moderate Wind SE    0
Moderate Wind SW    0
Moderate Wind NW    0
High Wind NE        0
High Wind SE        0
High Wind SW        0
High Wind NW        0
Year                0
dtype: int64


In [135]:
atlantic.head()

Name       Date Status Latitude Longitude  Maximum Wind  \
0              UNNAMED 1851-06-25     HU    28.0N     94.8W          80.0   
1              UNNAMED 1851-06-25     HU    28.0N     95.4W          80.0   
2              UNNAMED 1851-06-25     HU    28.0N     96.0W          80.0   
3              UNNAMED 1851-06-25     HU    28.1N     96.5W          80.0   
4              UNNAMED 1851-06-25     HU    28.2N     96.8W          80.0   

   Low Wind NE  Low Wind SE  Low Wind SW  Low Wind NW  Moderate Wind NE  \
0          NaN          NaN          NaN          NaN               NaN   
1          NaN          NaN          NaN          NaN               NaN   
2          NaN          NaN          NaN          NaN               NaN   
3          NaN          NaN          NaN          NaN               NaN   
4          NaN          NaN          NaN          NaN               NaN   

   Moderate Wind SE  Moderate Wind SW  Moderate Wind NW  High Wind NE  \
0               NaN               NaN               NaN           NaN   
1               NaN               NaN               NaN           NaN   
2               NaN               NaN               NaN           NaN   
3               NaN               NaN               NaN           NaN   
4               NaN               NaN               NaN           NaN   

   High Wind SE  High Wind SW  High Wind NW  Year  
0           NaN           NaN           NaN  1851  
1           NaN           NaN           NaN  1851  
2           NaN           NaN           NaN  1851  
3           NaN           NaN           NaN  1851  
4           NaN           NaN           NaN  1851

The naming converntion for hurricanes started in 1950. This also might indicate better data quality, so we will use hurricane data from 1950 to present day and sort hurricanes by name

In [118]:
named =[]

for date in atlantic['Date']:
    if date > 19500000:
        named.append(True)
    else: 
        named.append(False)
date_range = pd.Series(named)
atlantic_df = atlantic[date_range]

atlantic_df.head()

Name      Date Status Latitude Longitude  Maximum Wind  \
21879                 ABLE  19500812     TS    17.1N     55.5W            35   
21880                 ABLE  19500812     TS    17.7N     56.3W            40   
21881                 ABLE  19500812     TS    18.2N     57.4W            45   
21882                 ABLE  19500812     TS    19.0N     58.6W            50   
21883                 ABLE  19500813     TS    20.0N     60.0W            50   

       Low Wind NE  Low Wind SE  Low Wind SW  Low Wind NW  Moderate Wind NE  \
21879         -999         -999         -999         -999              -999   
21880         -999         -999         -999         -999              -999   
21881         -999         -999         -999         -999              -999   
21882         -999         -999         -999         -999              -999   
21883         -999         -999         -999         -999              -999   

       Moderate Wind SE  Moderate Wind SW  Moderate Wind NW  High Wind NE  \
21879              -999              -999              -999          -999   
21880              -999              -999              -999          -999   
21881              -999              -999              -999          -999   
21882              -999              -999              -999          -999   
21883              -999              -999              -999          -999   

       High Wind SE  High Wind SW  High Wind NW  
21879          -999          -999          -999  
21880          -999          -999          -999  
21881          -999          -999          -999  
21882          -999          -999          -999  
21883          -999          -999          -999

In [119]:
atlantic['Date'] = pd.to_datetime(atlantic['Date'], format='%Y%m%d').dt.date


In [120]:
atlantic.loc[0,'Date']

datetime.date(1851, 6, 25)

In [121]:
atlantic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49105 entries, 0 to 49104
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Name              49105 non-null  object
 1   Date              49105 non-null  object
 2   Status            49105 non-null  object
 3   Latitude          49105 non-null  object
 4   Longitude         49105 non-null  object
 5   Maximum Wind      49105 non-null  int64 
 6   Low Wind NE       49105 non-null  int64 
 7   Low Wind SE       49105 non-null  int64 
 8   Low Wind SW       49105 non-null  int64 
 9   Low Wind NW       49105 non-null  int64 
 10  Moderate Wind NE  49105 non-null  int64 
 11  Moderate Wind SE  49105 non-null  int64 
 12  Moderate Wind SW  49105 non-null  int64 
 13  Moderate Wind NW  49105 non-null  int64 
 14  High Wind NE      49105 non-null  int64 
 15  High Wind SE      49105 non-null  int64 
 16  High Wind SW      49105 non-null  int64 
 17  High Wind NW

In [122]:
#The Date column is still showing up as an object because '.dt.date' returns an object, not a pandas datetime object. So we must run 'pd.to_datetime' again

atlantic['Date'] = pd.to_datetime(atlantic['Date'])

In [123]:
atlantic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49105 entries, 0 to 49104
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Name              49105 non-null  object        
 1   Date              49105 non-null  datetime64[ns]
 2   Status            49105 non-null  object        
 3   Latitude          49105 non-null  object        
 4   Longitude         49105 non-null  object        
 5   Maximum Wind      49105 non-null  int64         
 6   Low Wind NE       49105 non-null  int64         
 7   Low Wind SE       49105 non-null  int64         
 8   Low Wind SW       49105 non-null  int64         
 9   Low Wind NW       49105 non-null  int64         
 10  Moderate Wind NE  49105 non-null  int64         
 11  Moderate Wind SE  49105 non-null  int64         
 12  Moderate Wind SW  49105 non-null  int64         
 13  Moderate Wind NW  49105 non-null  int64         
 14  High Wind NE      4910

Creating a new column for years

In [124]:
atlantic['Year'] = atlantic['Date'].dt.year
atlantic.head()

Name       Date Status Latitude Longitude  Maximum Wind  \
0              UNNAMED 1851-06-25     HU    28.0N     94.8W            80   
1              UNNAMED 1851-06-25     HU    28.0N     95.4W            80   
2              UNNAMED 1851-06-25     HU    28.0N     96.0W            80   
3              UNNAMED 1851-06-25     HU    28.1N     96.5W            80   
4              UNNAMED 1851-06-25     HU    28.2N     96.8W            80   

   Low Wind NE  Low Wind SE  Low Wind SW  Low Wind NW  Moderate Wind NE  \
0         -999         -999         -999         -999              -999   
1         -999         -999         -999         -999              -999   
2         -999         -999         -999         -999              -999   
3         -999         -999         -999         -999              -999   
4         -999         -999         -999         -999              -999   

   Moderate Wind SE  Moderate Wind SW  Moderate Wind NW  High Wind NE  \
0              -999              -999              -999          -999   
1              -999              -999              -999          -999   
2              -999              -999              -999          -999   
3              -999              -999              -999          -999   
4              -999              -999              -999          -999   

   High Wind SE  High Wind SW  High Wind NW  Year  
0          -999          -999          -999  1851  
1          -999          -999          -999  1851  
2          -999          -999          -999  1851  
3          -999          -999          -999  1851  
4          -999          -999          -999  1851

Dropping duplicate storms

In [125]:
atlantic_df = atlantic_df.sort_values(by='Maximum Wind', ascending=False)
atlantic_df = atlantic_df.drop_duplicates(subset='Name', keep="first")
atlantic_df.head()

Name      Date Status Latitude Longitude  Maximum Wind  \
33731                ALLEN  19800807     HU    21.8N     86.4W           165   
44435                WILMA  20051019     HU    17.3N     82.8W           160   
36445              GILBERT  19880914     HU    19.7N     83.8W           160   
40604                MITCH  19981027     HU    17.2N     83.8W           155   
44318                 RITA  20050922     HU    24.8N     87.6W           155   

       Low Wind NE  Low Wind SE  Low Wind SW  Low Wind NW  Moderate Wind NE  \
33731         -999         -999         -999         -999              -999   
44435          170          125           90          140                70   
36445         -999         -999         -999         -999              -999   
40604         -999         -999         -999         -999              -999   
44318          160          140          100          160               100   

       Moderate Wind SE  Moderate Wind SW  Moderate Wind NW  High Wind NE  \
33731              -999              -999              -999          -999   
44435                45                45                70            45   
36445              -999              -999              -999          -999   
40604              -999              -999              -999          -999   
44318                80                60                90            60   

       High Wind SE  High Wind SW  High Wind NW  
33731          -999          -999          -999  
44435            20            20            45  
36445          -999          -999          -999  
40604          -999          -999          -999  
44318            60            40            50

In [126]:
atlantic_df.shape

(288, 18)

In [127]:
atlantic_df = atlantic_df.sort_values(by="Date", ascending=True)
atlantic_df.head()

Name      Date Status Latitude Longitude  Maximum Wind  \
21910                 ABLE  19500819     HU    31.7N     75.5W           110   
22074                  DOG  19500905     HU    24.6N     67.7W           125   
22229               GEORGE  19501004     HU    34.2N     68.0W            95   
22269                 ITEM  19501010     HU    19.4N     95.4W            90   
22291                  JIG  19501015     HU    33.0N     59.7W           100   

       Low Wind NE  Low Wind SE  Low Wind SW  Low Wind NW  Moderate Wind NE  \
21910         -999         -999         -999         -999              -999   
22074         -999         -999         -999         -999              -999   
22229         -999         -999         -999         -999              -999   
22269         -999         -999         -999         -999              -999   
22291         -999         -999         -999         -999              -999   

       Moderate Wind SE  Moderate Wind SW  Moderate Wind NW  High Wind NE  \
21910              -999              -999              -999          -999   
22074              -999              -999              -999          -999   
22229              -999              -999              -999          -999   
22269              -999              -999              -999          -999   
22291              -999              -999              -999          -999   

       High Wind SE  High Wind SW  High Wind NW  
21910          -999          -999          -999  
22074          -999          -999          -999  
22229          -999          -999          -999  
22269          -999          -999          -999  
22291          -999          -999          -999